### ISSUE TITLE
Output from network inconsistent after save/load

### ISSUE THREAD

**Issue by zmonoid** [pytorch/pytorch#13242]
## 🐛 Bug

We fix the random seed, save and reload the network during training, but found the network output is different although the network parameters and input is exactly the same. See below code modified from pytorch example mnist. We insert save/load code during training:
```python
        if batch_idx == 55:
            torch.save({
                    'model': model.state_dict(),
                    }, 'save.pth')

            xmodel = Net()
            xdata = torch.load('save.pth')
            xmodel.load_state_dict(xdata['model'])
            xmodel.to(device)

            yhat = xmodel(data)

            model_conv1w_diff = np.abs(model.conv1.weight.cpu().detach().numpy() - xmodel.conv1.weight.cpu().detach().numpy()).max()
            model_fc2w_diff = np.abs(model.fc2.weight.cpu().detach().numpy() - xmodel.fc2.weight.cpu().detach().numpy()).max()
            model_fc2b_diff = np.abs(model.fc2.bias.cpu().detach().numpy() - xmodel.fc2.bias.cpu().detach().numpy()).max()

            y_diff = np.abs(output.cpu().detach().numpy() - yhat.cpu().detach().numpy()).max()
            print(model_conv1w_diff, model_fc2w_diff, model_fc2b_diff, y_diff)
```

The print output shows 
```
0.0 0.0 0.0 0.82663274
```

## To Reproduce
Run the following code
```python
from __future__ import print_function
import argparse
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
import numpy as np
import pdb

torch.manual_seed(7)

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 10, kernel_size=5)
        self.conv2 = nn.Conv2d(10, 20, kernel_size=5)
        self.conv2_drop = nn.Dropout2d()
        self.fc1 = nn.Linear(320, 50)
        self.fc2 = nn.Linear(50, 10)

    def forward(self, x):
        x = F.relu(F.max_pool2d(self.conv1(x), 2))
        x = F.relu(F.max_pool2d(self.conv2_drop(self.conv2(x)), 2))
        x = x.view(-1, 320)
        x = F.relu(self.fc1(x))
        x = F.dropout(x, training=self.training)
        x = self.fc2(x)
        return F.log_softmax(x, dim=1)

def train(args, model, device, train_loader, optimizer, epoch):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = data.to(device), target.to(device)
        data.requires_grad_()
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()

        if batch_idx == 55:
            torch.save({
                    'model': model.state_dict(),
                    }, 'save.pth')

            xmodel = Net()
            xdata = torch.load('save.pth')
            xmodel.load_state_dict(xdata['model'])
            xmodel.to(device)

            yhat = xmodel(data)

            model_conv1w_diff = np.abs(model.conv1.weight.cpu().detach().numpy() - xmodel.conv1.weight.cpu().detach().numpy()).max()
            model_fc2w_diff = np.abs(model.fc2.weight.cpu().detach().numpy() - xmodel.fc2.weight.cpu().detach().numpy()).max()
            model_fc2b_diff = np.abs(model.fc2.bias.cpu().detach().numpy() - xmodel.fc2.bias.cpu().detach().numpy()).max()

            y_diff = np.abs(output.cpu().detach().numpy() - yhat.cpu().detach().numpy()).max()
            print(model_conv1w_diff, model_fc2w_diff, model_fc2b_diff, y_diff)

            pdb.set_trace()

        optimizer.step()

        if batch_idx % args.log_interval == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                epoch, batch_idx * len(data), len(train_loader.dataset),
                100. * batch_idx / len(train_loader), loss.item()))

def test(args, model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item() # sum up batch loss
            pred = output.max(1, keepdim=True)[1] # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)
    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))

def main():
    # Training settings
    parser = argparse.ArgumentParser(description='PyTorch MNIST Example')
    parser.add_argument('--batch-size', type=int, default=64, metavar='N',
                        help='input batch size for training (default: 64)')
    parser.add_argument('--test-batch-size', type=int, default=1000, metavar='N',
                        help='input batch size for testing (default: 1000)')
    parser.add_argument('--epochs', type=int, default=10, metavar='N',
                        help='number of epochs to train (default: 10)')
    parser.add_argument('--lr', type=float, default=0.01, metavar='LR',
                        help='learning rate (default: 0.01)')
    parser.add_argument('--momentum', type=float, default=0.5, metavar='M',
                        help='SGD momentum (default: 0.5)')
    parser.add_argument('--no-cuda', action='store_true', default=False,
                        help='disables CUDA training')
    parser.add_argument('--seed', type=int, default=1, metavar='S',
                        help='random seed (default: 1)')
    parser.add_argument('--log-interval', type=int, default=10, metavar='N',
                        help='how many batches to wait before logging training status')
    args = parser.parse_args()
    use_cuda = not args.no_cuda and torch.cuda.is_available()

    torch.manual_seed(args.seed)

    device = torch.device("cuda" if use_cuda else "cpu")

    kwargs = {'num_workers': 1, 'pin_memory': True} if use_cuda else {}
    train_loader = torch.utils.data.DataLoader(
        datasets.MNIST('../data', train=True, download=True,
                       transform=transforms.Compose([
                           transforms.ToTensor(),
                           transforms.Normalize((0.1307,), (0.3081,))
                       ])),
        batch_size=args.batch_size, shuffle=True, **kwargs)
    test_loader = torch.utils.data.DataLoader(
        datasets.MNIST('../data', train=False, transform=transforms.Compose([
                           transforms.ToTensor(),
                           transforms.Normalize((0.1307,), (0.3081,))
                       ])),
        batch_size=args.test_batch_size, shuffle=True, **kwargs)


    model = Net().to(device)
    optimizer = optim.SGD(model.parameters(), lr=args.lr, momentum=args.momentum)

    for epoch in range(1, args.epochs + 1):
        train(args, model, device, train_loader, optimizer, epoch)
        test(args, model, device, test_loader)


if __name__ == '__main__':
    main()

```

## Expected behavior

Expected that printed output to be zero for all.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

```
Collecting environment information...
PyTorch version: 0.4.1.post2
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: version 3.10.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1070
Nvidia driver version: 390.48
cuDNN version: Probably one of the following:
/usr/local/MATLAB/R2018a/bin/glnxa64/libcudnn.so.7.0.3
/usr/local/cudnn7-cu90/lib64/libcudnn.so.7.2.1
/usr/local/cudnn7-cu90/lib64/libcudnn_static.a
/usr/local/cudnn7005-cu90/lib64/libcudnn.so.7.0.5
/usr/local/cudnn7005-cu90/lib64/libcudnn_static.a

Versions of relevant libraries:
[pip] Could not collect
[conda] torch                     0.4.1.post2               <pip>
[conda] torchvision               0.2.1                     <pip>

```

**Comment by Kaixhin** [comment]
Just reading the code quickly, you're testing two models in training mode with `nn.Dropout` on, so you'd want to check outputs after setting `model.eval()` and `xmodel.eval()`.

**Comment by zmonoid** [comment]
Yes, that is the issue.
Thanks.
